ESA_EHST_XMM: retrieve metadata and data from program, TAP: download in streaming, XMM: updated RMF matrices

jespinosaar opened the pull request:
Dear Astroquery team,

This PR aims to improve two modules:

- TAP: downloads are now executed in streaming mode. This will allow us to download large files without a huge memory consumption.
- eHST Module: we have included two new methods to retrieve observations and data associated to a specific program.
- XMM Module: RMF matrices updated to v21.

Thanks a lot in advance for your support!

Kind regards.
Javi
cc @esdc-esac-esa-int

pep8speaks commented:
Hello @jespinosaar! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:





* In the file [`astroquery/esa/hubble/tests/test_esa_hubble_remote.py`](https://github.com/astropy/astroquery/blob/576d47c4050901d34f390253240fe6b880290f7b/astroquery/esa/hubble/tests/test_esa_hubble_remote.py):

> [Line 133:1](https://github.com/astropy/astroquery/blob/576d47c4050901d34f390253240fe6b880290f7b/astroquery/esa/hubble/tests/test_esa_hubble_remote.py#L133): [W391](https://duckduckgo.com/?q=pep8%20W391) blank line at end of file

jespinosaar commented:
Great! it seems all the issues are now solved.

bsipocz commented on astroquery/esa/hubble/core.py:
```diff
@@ -215,6 +219,85 @@ def get_observation_type(self, observation_id):
             raise ValueError("Invalid Observation ID")
         return obs_type
 
+    def get_observations_from_program(self, program, *, output_file=None,
```
I wonder whether there is indeed any need for this convenience method at the price of complicating the API, as all it does is to pass on the parameters to `query_criteria` (maybe a well written example in the docs would be be sufficient)

bsipocz replied:
Thanks for the explanation, it sounds good.

bsipocz commented on astroquery/esa/hubble/core.py:
```diff
@@ -215,6 +219,85 @@ def get_observation_type(self, observation_id):
+
+        Returns
+        -------
+        None. It downloads the artifact indicated
```
As far as I see doesn't, it returns a Table

jespinosaar replied:
Indeed, it returns a table. Copy/paste problem! It has been fixed now.

bsipocz commented on astroquery/esa/hubble/core.py:
```diff
@@ -215,6 +219,85 @@ def get_observation_type(self, observation_id):
+        return self.query_criteria(proposal=program, output_file=output_file,
+                                   output_format=output_format, async_job=False, verbose=verbose)
+
+    def download_files_from_program(self, program, *, folder=os.getcwd(), calibration_level=None,
```
What about using a more generic name, such as `download_files` or `download_products` and making `program` optional? (e.g. allowing the download of any results of `query_criteria`, as I see the only difference in the signature is that this mandates `program` to be specified, while it's optional for `query_criteria`)

jespinosaar replied:
Same as before, the wanted an explicit method. If it is not ok with you, please let me know.

bsipocz replied:
Overall I think it's better to try not to have too many methods that only differ minimally and use more parameters if needed (good for removing code duplications, etc.). However, it's always difficult to say where the line is especially if there is already a user request for a functionality.
So it's all good for now.

bsipocz commented on astroquery/esa/hubble/core.py:
```diff
@@ -239,7 +322,7 @@ def _get_product_filename(self, product_type, filename):
             return f"{filename}.zip"
 
-    def get_artifact(self, artifact_id, *, filename=None, verbose=False):
+    def get_artifact(self, artifact_id, *, filename=None, folder=os.getcwd(), verbose=False):
```
same here (and everywhere else below, too, maybe use `None` as default that defaults to current dir

jespinosaar replied:
Removed and replaced with None.

bsipocz commented on astroquery/esa/jwst/core.py:
```diff
@@ -39,6 +39,21 @@
 __all__ = ['Jwst', 'JwstClass']
 
 
+def _check_rename_to_gz(filename):
```
I wonder whether it would make sense to have some esa utilities somewhere that lets some shared functions to be reused rather than repeated? (certainly not required for this PR, I'm just wondering about it). It could live either under `esa/utils` or `utils/esa`. I suppose this, and some of the test mock classes can be dumped in there for sure, but maybe there is more.

jespinosaar replied:
I have removed this from ESA HST and JWST modules and created a new one under esa/utils. It is included in my latest commits.

bsipocz commented on docs/esa/hubble/hubble.rst:
```diff
@@ -545,6 +545,54 @@ returns the corresponding HAP or HST observation
+  >>> from astroquery.esa.hubble import ESAHubble
+  >>> esahubble = ESAHubble()
+  >>> esahubble.get_observations_from_program(program=5773)
+  <Table length=68>
+  band_name calibration_level collection ...      wave_max      wave_min
+  object        int32         str64    ...      float64       float64
+  --------- ----------------- ---------- ... ------------------ --------
+  Optical                 2        HST ... 502.58000000000004   499.89
+  Optical                 2        HST ...  570.9399999999999   522.62
+  Optical                 2        HST ...             675.58   670.86
+  Optical                 2        HST ...             657.46    655.3
+  Optical                 2        HST ...  570.9399999999999   522.62
+  Optical                 1        HST ... 502.58000000000004   499.89
+  Optical                 2        HST ...             657.46    655.3
+  Optical                 1        HST ...             657.46    655.3
+  Optical                 2        HST ... 502.58000000000004   499.89
+  Optical                 1        HST ...             675.58   670.86
+      ...               ...        ... ...                ...      ...
+  Optical                 3        HLA ...             570.94   522.62
+  Optical                 3        HLA ...             570.94   522.62
+  Optical                 3        HLA ...             570.94   522.62
+  Optical                 3        HLA ...             570.94   522.62
+  Optical                 3        HLA ... 502.58000000000004   499.89
+  Optical                 3        HLA ... 502.58000000000004   499.89
+  Optical                 3        HLA ... 502.58000000000004   499.89
+  Optical                 3        HLA ... 502.58000000000004   499.89
+  Optical                 3        HLA ... 502.58000000000004   499.89
+  Optical                 3        HLA ... 502.58000000000004   499.89
```
whitespace is off here failing this test.

Btw, the latest version of `pytest-doctestplus` has a new feature to auto-update the doctests, you may find it useful for cases like this: 
command: `pytest docs/esa/hubble/hubble.rst --doctest-plus-generate-diff=overwrite -R`
more info: https://github.com/scientific-python/pytest-doctestplus/blob/main/README.rst#fixing-existing-docstrings

```suggestion
  <Table length=68>
  band_name calibration_level collection ...      wave_max      wave_min
    object        int32         str64    ...      float64       float64
  --------- ----------------- ---------- ... ------------------ --------
    Optical                 2        HST ... 502.58000000000004   499.89
    Optical                 1        HST ...             657.46    655.3
    Optical                 2        HST ...             657.46    655.3
    Optical                 2        HST ...             675.58   670.86
    Optical                 2        HST ... 502.58000000000004   499.89
    Optical                 2        HST ... 502.58000000000004   499.89
    Optical                 2        HST ...  570.9399999999999   522.62
    Optical                 1        HST ...  570.9399999999999   522.62
    Optical                 2        HST ...             657.46    655.3
    Optical                 2        HST ...             675.58   670.86
        ...               ...        ... ...                ...      ...
    Optical                 3        HLA ...             570.94   522.62
    Optical                 3        HLA ...             570.94   522.62
    Optical                 3        HLA ...             570.94   522.62
    Optical                 3        HLA ...             570.94   522.62
    Optical                 3        HLA ... 502.58000000000004   499.89
    Optical                 3        HLA ... 502.58000000000004   499.89
    Optical                 3        HLA ... 502.58000000000004   499.89
    Optical                 3        HLA ... 502.58000000000004   499.89
    Optical                 3        HLA ... 502.58000000000004   499.89
    Optical                 3        HLA ... 502.58000000000004   499.89
```

jespinosaar replied:
I finally removed it, as the result may change.

jespinosaar replied:
And many thanks for the command, for sure we will use it.

bsipocz reviewed (commented):
Thank you @jespinosaar! 

I left a few comments, but most of them are meant to be a discussion starter rather than a blocker for merging, so please do share your insights. My overall motivation is to eventually converge towards using the same method names throughout the modules for the very similar functionalities, and to avoid over-complicating the API with too many methods that are in fact just calling one other method without much else. But each archive is different enogh that this desire sometimes doesn't make sense.

jespinosaar commented on astroquery/esa/hubble/core.py:
```diff
@@ -63,7 +64,7 @@ def __init__(self, *, tap_handler=None, show_messages=True):
 
     def download_product(self, observation_id, *, calibration_level=None,
-                         filename=None, verbose=False, product_type=None):
+                         filename=None, folder=os.getcwd(), verbose=False, product_type=None):
```
I have replaced os.getcwd with None. For the output_folders, I created a new method so, in case folder is None, only the filename is returned, so it is downloaded in the current directory. If it is provided, then using os.path we get the absolute path.

bsipocz replied:
I'm afraid the change hasn't been committed for this method

jespinosaar commented on CHANGES.rst:
```diff
@@ -56,6 +56,11 @@ esa.hubble
 - Status and maintenance messages from eHST TAP when the module is instantiated. get_status_messages method to retrieve them. [#2597]
 - New methods to download single files ``download_file`` and download FITS associated to an observation ``download_fits_files``. [#2797]
 - New function to retrieve all the files associated to an observation. [#2797]
+- New methods to retrieve metadata and files associated to a proposal. [#2910]
```
New method names included.

jespinosaar commented:
Many thanks for your inputs @bsipocz!

I have implemented all the changes and answered all the comments. The objective of the new methods is to offer to our users a explicit method to download data from a program. It is true this can be achieved using other methods, but we wanted to make it easier for users to find it (and they even requested it). But please, let me know if you don't agree with this approach and we will iterate.

Please let me know if further changes are required.

Kind regards,
Javi

P.D. happy new year!

jespinosaar commented:
Hi @bsipocz , is it ok with you if we keep the new methods? Thanks!

jespinosaar commented:
Hi all, we need to plan if more effort is required on this PR. Could you please let us know if more changes are required? Thanks in advance!

bsipocz commented:
> Hi all, we need to plan if more effort is required on this PR.

Sorry for being slow on coming back for the review, it wasn't held up on purpose.

bsipocz reviewed (approved):
This looks good now with one super minor oversight.

Thanks!

jespinosaar commented:
> This looks good now with one super minor oversight.
> 
> Thanks!

Great! I have fixed it and updated the PR (there was another remote test that was failing, due to changes in our DB).

jespinosaar commented:
> > Hi all, we need to plan if more effort is required on this PR.
> 
> Sorry for being slow on coming back for the review, it wasn't held up on purpose.

Please do not worry! I was just planning our next sprint. And thanks for looking into this.

jespinosaar commented:
As always, many thanks @bsipocz !!!

bsipocz commented:
Yeap, I can do that. And I really need to wrap up a few things and get 0.4.7 out of the door and then switch to a better versioning with more frequent releases.

bsipocz commented:
https://pypi.org/project/astroquery/0.4.7.dev9064/
